node/beacon-chain-handler/src/importer.rs

```diff
@@ -91,44 +90,46 @@ impl BlockImporter {
             parent_block_hash: parent_hash,
             shard_id: shard_block.body.header.shard_id,
         };
-        let mut apply_result = self.runtime.write().apply(
+        let apply_result = self.runtime.write().check(
             &apply_state,
             &prev_shard_block.body.new_receipts,
-            shard_block.body.transactions.clone()
+            &shard_block.body.transactions
         );
-        if apply_result.root != prev_shard_header.body.merkle_root_state {
-            info!(
-                "Merkle root {} is not equal to received {} after applying the transactions from {:?}",
-                prev_shard_header.body.merkle_root_state,
-                apply_result.root,
-                beacon_block
-            );
-            return;
-        }
-        if apply_result.filtered_transactions.len() != num_transactions {
-            info!(
-                "Imported block has transactions that were filtered out while merkle roots match in {:?}",
-                beacon_block
-            );
-            return;
+        match apply_result {
+            Some((mut transaction, root)) => {
+                if root != prev_shard_header.body.merkle_root_state {
+                    info!(
+                        "Merkle root {} is not equal to received {} after applying the transactions from {:?}",
+                        prev_shard_header.body.merkle_root_state,
+                        root,
+                        beacon_block
+                    );
+                    return;
+                }
+                self.state_db.commit(&mut transaction).ok();
+                self.shard_chain.insert_block(shard_block);
+                self.beacon_chain.insert_block(beacon_block);
+            }
+            None => {
+                info!(
+                    "Imported block has transactions that were filtered out while merkle roots match in {:?}",
```

Comment:
"Found incorrect transactions in the block" (filtered would be weird for anybody who reads this, and this condition doesn't actually mean merkle roots match)